runtime/spl/array_iterator.h

```diff
@@ -0,0 +1,62 @@
+// Compiler for PHP (aka KPHP)
+// Copyright (c) 2022 LLC «V Kontakte»
+// Distributed under the GPL v3 License, see LICENSE.notice.txt
+
+#pragma once
+
+#include "common/algorithms/hashes.h"
+#include "common/wrappers/string_view.h"
+
+#include "runtime/dummy-visitor-methods.h"
+#include "runtime/kphp_core.h"
+#include "runtime/refcountable_php_classes.h"
+
+// C$ArrayIterator implements SPL ArrayIterator class.
+struct C$ArrayIterator : public refcountable_php_classes<C$ArrayIterator>, private DummyVisitorMethods {
+  // we store an array to keep a living reference to it while iterator is valid;
+  // also we may implement rewind() method later if we feel like it
+  array<mixed> arr;
+  array<mixed>::const_iterator it;
+  array<mixed>::const_iterator end;
```

Comment:
Does PHP version of ArrayIterator work when an array is modified?
This version does not, since modifying arr could leave these iterators stale.

Reply:
There are tests that check this behavior.
ArrayIterator iterates over the provided array without reference.
So if that array is modified externally, we'll end up iterating over a copy that is held by the ArrayIterator.

Reply:
I see, thank you